### PR TITLE
Introduce Mark Mode

### LINKED ISSUE #715: Terminal: Support keyboard text selection & navigation
Hello,

I'm using tmux on Linux machines and ConEmu on Windows machines and I really love that I can do text selection with the keyboard.

In ConEmu, there is a keybinding for "Start text selection (like text editors)", whereas in tmux there is the same thing with `<prefix><b>` (I think that is default), where I get a cursor that I can move with the arrow keys around the console window.
While doing this, I can select text while holding shift and (in case of ConEmu) press ENTER to copy the text into the clipboard and end the text selection.
It can also be compared with Vim's visual mode.

It would be nice if the new Windows Terminal could support this out of the box, so I can use it in e.g. PowerShell :-)

### PR DESCRIPTION
**Pull request by carlos-zamora**
## Summary of the Pull Request
Introduces a non-configurable version of mark mode to Windows Terminal. It has the following interactions defined:
- <kbd>ctrl+shift+m</kbd> --> Enter Mark Mode
- when in Mark Mode...
	- <kbd>ESC</kbd> --> Exit Mark Mode
	- arrow keys --> move "start"
	- <kbd>shift</kbd> + arrow keys --> anchor "start", move "end"
	- <kbd>ctrl+a</kbd> --> select all
- when a selection is active...

When in mark mode, the cursor does not blink.

## References
#4993 - [Epic] Keyboard Selection

## PR Checklist
* [X] Closes #715
* [X] Provides a resolution for #11985

## Detailed Description of the Pull Request / Additional comments
- `TermControl`:
	- `TermControl.cpp` just adds logic to prevent the cursor from blinking when in mark mode
- `ControlCore`
	- in the same place we handle quick edit, we add an entry point to mark mode
- `TerminalCore`
	- this leverages `UpdateSelection()` and other quick edit functions to make mark mode happen

## Validation Steps Performed
- [x] Make selection, split pane, close pane
	- NOTE: A similar scenario caused a crash at one point. Really weird. Keep an eye on it.
- [x] Cursor is off when in mark mode
- [x] general movement/selection
- [x] general movement/selection that forces the viewport to move
- [x] In mark mode, selectAll...
	- [x] arrow keys --> move start
	- [x] shift + arrow keys --> move end
- [x] (regardless of mark mode) if selection active, enter --> copy to clipboard

### CODE REVIEW

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 1]
```diff
@@ -387,33 +387,56 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
-        if (HasSelection() &&
-            !KeyEvent::IsModifierKey(vkey) &&
+        if (!KeyEvent::IsModifierKey(vkey) &&
             vkey != VK_SNAPSHOT &&
             keyDown)
         {
-            // try to update the selection
-            if (const auto updateSlnParams{ ::Terminal::ConvertKeyEventToUpdateSelectionParams(modifiers, vkey) })
+            // TODO CARLOS: should we make this ctrl+shift+m instead?
+            if (modifiers.IsCtrlPressed() && vkey == 'M')
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:395]: Whatcha think?

Comment (zadjii-msft) [reply src/cascadia/TerminalControl/ControlCore.cpp:395]: make this an action, default it to ctrl+shift+m. The keybindings w/in mark mode don't need to be configurable now, but the toggle should be.

File: src/cascadia/TerminalCore/TerminalSelection.cpp [hunk 2]
```diff
@@ -281,11 +299,17 @@ Terminal::UpdateSelectionParams Terminal::ConvertKeyEventToUpdateSelectionParams
...
+// Arguments:
+// - direction: the direction to move the selection endpoint in
+// - mode: the type of movement to be performed (i.e. move by word)
+// - (optional) moveSelectionEnd: used by mark mode. Can be used to explicitly move a specific endpoint. If true, move the "end" endpoint. Otherwise, move the start endpoint.
+void Terminal::UpdateSelection(SelectionDirection direction, SelectionExpansion mode, std::optional<bool> moveSelectionEnd)
 {
     // 1. Figure out which endpoint to update
     // One of the endpoints is the pivot, signifying that the other endpoint is the one we want to move.
-    const auto movingEnd{ _selection->start == _selection->pivot };
+    const auto movingEnd{ (_markMode && moveSelectionEnd.has_value()) ? *moveSelectionEnd : _selection->start == _selection->pivot };
```
Comment (zadjii-msft) [inline src/cascadia/TerminalCore/TerminalSelection.cpp:312]: For some reason I'm having a hard time parsing this `moveSelectionEnd` logic here. So, there's three cases:
* `nullopt`: not in mark mode. use `_selection->start == _selection->pivot`. We'll later just move the end anchor of the selection.
* `true`: in mark mode. Shift was pressed. we're just going to move the end anchor?
* `false`: in mark mode. We're going to move the start. But then later we're in mark mode, so we're just going to move the end to where the start is?

idk something about this just felt weird.

**Review by zadjii-msft — CHANGES_REQUESTED** [review]
NAK on magic `ctrl+m`

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 3]
```diff
@@ -1123,7 +1123,7 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         {
             // Manually show the cursor when a key is pressed. Restarting
             // the timer prevents flickering.
-            _core.CursorOn(true);
+            _core.CursorOn(!_core.IsInMarkMode());
```
[the remark this answers is not in the document]
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControl.cpp:1126]: Yeah, that's intentional (but definitely up for discussion on the design, actually). I'll just rapid fire random thoughts in my head for why I feel this is the desired behavior...
- the cursor is a bit distracting when it's blinking if you're in mark mode
- preventing the cursor from blinking makes it feel more clear that you're in a different mode (though we're looking to add the markers in #10865 )
- when in mark mode, the "cursor" should really be the selection marker (just like in a text editor). So there's no reason to show the actual cursor.
	- [Accessibility] we do need to track the cursor movements for some things. Maybe we should _actually_ be moving the cursor along with the selection? Idk. I'll have to test this out.

**Comment by zadjii-msft** [comment]
Team discuss: 

* [ ] yea make mark mode an action
* [ ] keys as-is is good, if someone really wants rebinding, we'll let them ask for it.
* [ ] make a follow up for toggle endpoint as an action - having a separate mode for "copy mode" is silly
* [ ] #shipit
* [ ] The docs have updateSelection actions in them, probably remove those
* [ ] Add a new top-level docs section to describe all this interaction

File: src/cascadia/TerminalCore/TerminalSelection.cpp (hunk 2 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalCore/TerminalSelection.cpp:312]: Ah, let me try and explain it this way...
- `nullopt` <-- we're just modifying an existing selection with the keyboard. So like, you created a selection with a mouse, and now you're just using <kbd>shift+arrows</kbd> to update it. Codeflow requires you to pivot around the "start" if you cross the endpoints.
- `true` <-- you are _exclusively_ only moving the "end". This is like when you're creating a selection in a text editor. "start" is anchored to the cursor, and "end" is moved; thus creating a selection.
- `false` <-- you are moving _both_ "start" and "end" together. This is when you're just moving the cursor in a text editor.

File: doc/cascadia/profiles.schema.json [hunk 4]
```diff
@@ -327,6 +328,7 @@
         "sendInput",
         "setColorScheme",
         "setTabColor",
+        "selectAll",
```
Comment (carlos-zamora) [inline doc/cascadia/profiles.schema.json:331]: _Technically_, this should be its own PR and cherry-picked into the 1.14 branch... but the schema is online so I really don't see a reason for that.

File: src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw [hunk 5]
```diff
@@ -525,4 +525,7 @@
   <data name="SelectAllCommandKey" xml:space="preserve">
     <value>Select all text</value>
   </data>
+  <data name="MarkModeCommandKey" xml:space="preserve">
+    <value>Toggle Mark Mode</value>
```
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/Resources/en-US/Resources.resw:529]: I should leave a comment here for the translators.

File: doc/cascadia/profiles.schema.json (hunk 4 above)
Comment (DHowett) [reply doc/cascadia/profiles.schema.json:331]: The schema is an aka.ms link pointed to the stable version's schema. If we move that forward to 1.14, the schema will report that selectAll is invalid...

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 6]
```diff
@@ -392,11 +392,25 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+                auto lock = _terminal->LockForWriting();
+                _terminal->SelectAll();
+                _renderer->TriggerSelection();
+                return true;
+            }
+            else if (vkey == VK_RETURN)
+            {
+                CopySelectionToClipboard(false, nullptr);
+                return true;
+            }
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/ControlCore.cpp:406]: You know what... I should remove this. The user can bind "enter" to "copy" if they want this behavior.

File: src/cascadia/TerminalCore/TerminalSelection.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalCore/TerminalSelection.cpp:322]: that's a good description - mind adding that in the comments? That would make this clearer at least, to know what it is trying to do.

**Review by zadjii-msft — APPROVED** [review]

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 7]
```diff
@@ -392,11 +392,20 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+
             // try to update the selection
-            if (const auto updateSlnParams{ ::Terminal::ConvertKeyEventToUpdateSelectionParams(modifiers, vkey) })
+            if (const auto updateSlnParams{ _terminal->ConvertKeyEventToUpdateSelectionParams(modifiers, vkey) })
             {
                 auto lock = _terminal->LockForWriting();
-                _terminal->UpdateSelection(updateSlnParams->first, updateSlnParams->second);
+                const std::optional<bool> markModeOverride = _terminal->IsInMarkMode() ? modifiers.IsShiftPressed() : std::optional<bool>{};
+                _terminal->UpdateSelection(updateSlnParams->first, updateSlnParams->second, markModeOverride);
```
Comment (lhecker) [inline src/cascadia/TerminalControl/ControlCore.cpp:408]: The `UpdateSelection` logic already depends on `_markMode` and now it depends on it twice - once explicitly internally (`_markMode`) and once implicitly (this parameter being dependent on `_markMode`, but computed from "the outside").

You can add a `ControlKeyStates mods` parameter instead and pass `modifiers` directly, the same way you do it with `ConvertKeyEventToUpdateSelectionParams`. Then ignore the parameter depending on the `_markMode` state.

This works well IMO since `ConvertKeyEventToUpdateSelectionParams` already has a very closely related `(_markMode || mods.IsShiftPressed())` condition.

**Review by lhecker — APPROVED** [review]

**Review by DHowett — CHANGES_REQUESTED** [review]
blocking for time to look today

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 8]
```diff
@@ -392,11 +392,19 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             vkey != VK_SNAPSHOT &&
             keyDown)
         {
+            if (modifiers.IsCtrlPressed() && vkey == 'A')
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:395]: I think this will impact the way I use my terminal negatively -- _any selection_ will cause `^A` to select all, but I have been known to select text and use `^A` to move to the beginning of the line (emacs key bindings).

Should this be behind a mark mode check?

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 9]
```diff
@@ -392,11 +392,19 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             vkey != VK_SNAPSHOT &&
             keyDown)
         {
+            if (modifiers.IsCtrlPressed() && vkey == 'A')
+            {
+                auto lock = _terminal->LockForWriting();
+                _terminal->SelectAll();
+                _renderer->TriggerSelection();
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:399]: I'm surprised we don't have a helper for this!

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/ControlCore.cpp:399]: yeah, I think at one point I made (or witnessed) the decision that triggering the renderer and (un)locking would be handled at the control core layer consistently to ensure we don't deadlock on accident. It's a little annoying, but we could swing the pendulum the other way hahaha

File: src/cascadia/TerminalControl/ControlCore.cpp (hunk 8 above)
Comment (carlos-zamora) [reply src/cascadia/TerminalControl/ControlCore.cpp]: added a mark mode check. I think I had that at one point and got it mixed up. Whoops!

I remember some discussion around it in the selectAll PR, but too lazy to bring that up rn.

**Review by DHowett — APPROVED** [review]

**Comment by ofek** [comment]
When will this be released? Also since there's not much documentation, can you please tell me if this will satisfy the use case I described here? https://github.com/microsoft/terminal/issues/6528

**Comment by ofek** [comment]
or is this the final feature? https://github.com/microsoft/terminal/issues/11868

**Comment by zadjii-msft** [comment]
@ofek We're planning on shipping this in Preview 1.15, the next Terminal release. Typically, these releases are about 6 weeks apart. We'll get to elaborating on the docs once we get closer to release.

Trying to parse everything there and translate it to what we've got so far:
* "click and drag (highlight) somewhere with the mouse"->"press any arrow key while holding shift"
  - I think this actually shipped in 1.13, IIRC. Definitely in 1.14.
* "At this point you can release the left-click (1) or shift (2). ... and you can now move anywhere with just the arrow keys"
  - @carlos-zamora would have to clarify this one for me. I _think_ that at this point, just arrow keys won't move the cursor. 
  - HOWEVER, if you started the selection by pressing `ctrl+shift+m` (or whatever you bind mark mode to), then just arrow keys _will_ move the cursor.
* "To highlight you then hold shift while using arrow keys similar to the highlighting logic of holding left-click while moving the mouse."
  - Yep, that's how mark mode works.
* "Copying can happen in 2 ways:...."
  - This I _don't_ think we have today (again, Carlos correct me if I'm wrong). We may need to add an arg to `copyText` to specifically ask that the selection is not dismissed.

**Comment by ofek** [comment]
Hey! So I spent this weekend learning & configuring the terminal. It's amazing, and I think I can immediately switch over when this is released. I wanted to test to be sure though so I downloaded the `drop` artifact from the CI like in https://dev.azure.com/ms/terminal/_build/results?buildId=322645&view=artifacts&pathAsName=false&type=publishedArtifacts

The `.\Release\x64\test\wt.exe` doesn't do anything and the `.\CascadiaPackage_0.0.1.0_x64.msix` requires a certificate to install. Is building from source required to test development versions?

**Comment by ofek** [comment]
I'm now able to use Windows Terminal without fatigue from heavy mouse movement. Thank you very much!!!
